### PR TITLE
Add Suffix newtype and use in PPSuffixHandler

### PR DESCRIPTION
**Pull request by sheaf**
This commit adds a `Suffix` newtype to describe suffixes as handled by suffix handlers & preprocessors, and changes the `PPSuffixHandler` type synonym to use it.

It also moves some type definitions from `Distribution.Simple.PreProcess` to the new module `Distribution.Simple.PreProcess.Types`.

As this commit changes the definition of `PPSuffixHandler`, it will break custom Setup scripts which use the `hookedPreProcessors` functionality. As such, this PR should at least make mention of the breakage in a changelog; but as there isn't all that much payoff to the change it might be best to hold off entirely until we can get rid of Custom setups.

### CODE REVIEW

**Comment by andreabedini** [comment]
What is the point of a newtype wrapper if we export the constructor?

If the breaking change is
```diff
type PPSuffixHandler =
-  (String, BuildInfo -> LocalBuildInfo -> ComponentLocalBuildInfo -> PreProcessor)
+  (Suffix, BuildInfo -> LocalBuildInfo -> ComponentLocalBuildInfo -> PreProcessor)
```
I would expect `Suffix` to be significantly different from a String.

Actually, maybe the breaking change can be avoided entirely, the field
```
hookedPreProcessors :: [PPSuffixHandler]
```
gets passed to Cabal. If I am not mistaken, all we would need to do to keep compatibility is converting that `String` to a `Suffix` as we receive it.

**Comment by sheaf** [comment]
The intent was more of documentation than anything else:

  - using a different type rather than `String` makes it a lot easier to understand type signatures, especially those which have other `String` arguments,
  - using a newtype rather than a type synonym makes it so that one is less likely to make a silly mistake, and can help enforce the invariant that a suffix is something like `"hs"` and not `".hs"`.

However, I don't particular mind making it a type synonym if that's what we need to do to avoid breaking people's custom Setup scripts.

NB: I have a few additional changes involving `Suffix` that I hope to push to this MR.

**Comment by sheaf** [comment]
I would of course like to use a different type than `String` (which IMO should basically never be used), but `Cabal` needs to change the representation of filepaths that it uses for that change to make sense.

**Comment by andreabedini** [comment]
Ok, I am ok with a newtype. The breaking change is minimal after all.

**Comment by sheaf** [comment]
> Do we have a build failure report for one of our test suites hitting this type mismatch already?

Yes, `AutogenModulesToggling` and `CustomPreProcess` ran into type-checking failures, which I solved without introducing CPP by enabling `OverloadedStrings`.

File: Cabal/src/Distribution/Simple/PreProcess/Types.hs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      -> Verbosity -- verbosity
+      -> IO () -- Should exit if the preprocessor fails
+  }
+
+-- | A suffix (or file extension).
+--
+-- Mostly used to decide which preprocessor to use, e.g. files with suffix @"y"@
+-- are usually processed by the @"happy"@ build tool.
+newtype Suffix = Suffix String
+  deriving (Eq, Ord, Show, Generic, IsString)
```
Comment (Kleidukos) [inline Cabal/src/Distribution/Simple/PreProcess/Types.hs:100]: Looking at CI, it looks like the stock Show instance of Suffix is used to print human-readable messages, which is not what it should be used for. Either Show should be derived `via String` or another function that unwraps it should be used to build this message:  `Could not find module: MyLibHelperModule with any suffix: ["gc","chs","hsc","x","y","ly","cpphs","hs","lhs","hsig","lhsig"].`

**Review by Kleidukos — CHANGES_REQUESTED** [review]

**Review by geekosaur — APPROVED** [review]

File: Cabal/src/Distribution/Simple/PreProcess/Types.hs (hunk 1 above)
Comment (andreabedini) [reply Cabal/src/Distribution/Simple/PreProcess/Types.hs:100]: Use
```
instance Pretty Suffix where
  pretty (Suffix s) = Disp.text s
```
and change `Cabal/src/Distribution/Simple/Errors.hs` to do
```
      ++ intercalate "," (map prettyShow suffixes)
```

Comment (alt-romes) [reply Cabal/src/Distribution/Simple/PreProcess/Types.hs:100]: Indeed. Fixing...

**Comment by Mikolaj** [comment]
@Kleidukos: I know you are super-busy. Shall I dismiss your blocking review, optimistically assuming all concerns have been addressed?

**Review by Kleidukos — APPROVED** [review]
Screenshot_20240222_131436

**Comment by alt-romes** [comment]
I addressed @Kleidukos' review.

**Comment by Kleidukos** [comment]
@sheaf any reason why you closed the PR?

**Comment by sheaf** [comment]
This PR was already merged when merging #9543.

**Comment by Kleidukos** [comment]
ok-thumbs-up

**Pull request closed without merging**